src/Cardano/Wallet/Api/Types/WalletState.hs

```diff
@@ -0,0 +1,62 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE LambdaCase #-}
+
+module Cardano.Wallet.Api.Types.WalletState
+    ( WalletState (..)
+    , WalletStateError (..)
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.Api.JSON
+    ( defaultRecordTypeOptions )
+import Data.Aeson
+    ( FromJSON (..), ToJSON (..), genericParseJSON, genericToJSON )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Wallet.Api.Types.Percentage as T
+import qualified Cardano.Wallet.Api.Types.WalletStateStatus as T
+
+data WalletState
+    = Ready
+    | Restoring T.Percentage
+    deriving (Eq, Generic, Show)
+
+instance FromJSON WalletState where
+    parseJSON x = either (fail . show) pure . validate =<< parseJSON x
+instance ToJSON WalletState where
+    toJSON = toJSON . unvalidate
+
+data UnvalidatedWalletState = UnvalidatedWalletState
+    { _status :: T.WalletStateStatus
+    , _progress :: Maybe T.Percentage
+    } deriving Generic
+
+data WalletStateError
+    = WalletInReadyStateCannotHaveProgressPercentage
+    | WalletInRestoringStateMustHaveProgressPercentage
+    deriving Show
+
+instance FromJSON UnvalidatedWalletState where
+    parseJSON = genericParseJSON defaultRecordTypeOptions
+instance ToJSON UnvalidatedWalletState where
+    toJSON = genericToJSON defaultRecordTypeOptions
+
+validate :: UnvalidatedWalletState -> Either WalletStateError WalletState
```

Comment:
Is this validation needed? I think the user can POST or PUT a wallet, but they can't update the `state` property.

Reply:
We might write clients in Haskell (or write tests of round-trip functionality). In such a case, this functionality would used during deserialization on the client end.

Of course, if this isn't needed, then we could also remove the `FromJSON` instance.

Reply:
OK, yes I didn't consider clients yet.